### PR TITLE
Split `updateParams` methods in tactics into `updateWorldParams` and `updateControlParams`

### LINKED ISSUE #775: Split the `updateParams` Methods On Tactics Into `updateWorldParams` and `updateControlParams`
### Description of the task
Split the `updateParams` methods in all tactics into `updateWorldParams` and `updateControlParams`. 
- NOTE: naming is flexible here, names in this issue are just suggestions
- `updateWorldParams` should take all the parameters required by the tactic that can be derived from the world (ex. `Field`, `Ball`, etc.)
- `updateControlParams` should take all other params (`kick_speed`, `pivot_radius`, etc.)
- please consider #776 when splitting the functions. Eventually `updateWorldParams` will be called from outside each `Play` to make plays simpler to write.

### Acceptance criteria

- [x] Split the `updateParams` methods for every tactic into `updateWorldParams` and `updateControlParams`
- [x] Change all usages of `updateParams` to call both methods in appropriate order

### PR DESCRIPTION
**Pull request by jsn0716**
### Description
Splits `updateParams` method in all tactics into two separate methods, `updateWorldParams` and `updateControlParams`. The method `updateWorldParams` updates parameters that can be derived from `World`, and `updateControlParams` updates all others.

To be more specific about what parameters were chosen to be updated in `updateWorldParams`, all parameters that were always derived from `World` outside of tests were included, such as:
- Parameters directly in `World` such as `Field` and `Ball`
- `Passing::Pass pass` in `PasserTactic` and `ReceiverTactic`
- `std::optional<Evaluation::EnemyThreat> enemy_threat` in `ShadowEnemyTactic`
- `std::optional<Point> chip_target` in `ShootGoalTactic`

This PR also removes `updateParams` method from `StopTestTactic` and `StopTactic` that were previously not updating any parameters.

### Testing Done
Since no new functions were added in this PR, testing was done by building and running existing tests. Tests should most likely be modified/added when the `TacticUpdateVisitor` from #776 is implemented.

### Resolved Issues
Resolves #775.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by jsn0716** [comment]
Just realized this PR is over the 500 lines limit at 561; most of the lines changed are just method name changes so it seems redundant to separate the pull request, but will do if requested.

**Comment by garethellis0** [comment]
@jonathanlew @akhilveeraghanta - please review!

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Can you take a look at the few areas I highlighted? I think they're more control params, but let me know what you think

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
This is much easier to follow, nice work!

 Just wondering how we are defining a "world" param vs a "control" param? Is a world param anything that is a part of the world obj?  Lets document the decision somewhere in this PR, (maybe tactic.h) so we be consistent from here on

**Comment by garethellis0** [comment]
@akhilveeraghanta @jsn0716 - Just looked over the issue for this and realized I missed something when defining the difference between "WorldParam" and "ControlParam". Basically there are some params that the Play needs to constantly update (like chip target, enemy threat, etc.). While some of these can technically be derived solely from the World (and hence matching my original definition of a "WorldParam"), they can't be derived from the world _independent of the play_. 

Sorry for the confusion, I'll go through and note any ones we should move!

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Sorry again for the confusion on this! I just reviewed the header files for now, didn't want to leave too many comments if stuff is gonna get re-distributed between `getWorldParams` and `getControlParams` anyhow.

Some of my comments might be a bit tricky to understand if you're not super familiar with the Play-Tactic architecture. Please feel free to comment reply or DM me on slack if you've got any questions at all!

File: src/software/ai/hl/stp/tactic/patrol_tactic.h
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/tactic/patrol_tactic.h]: Maybe it can be moved to the constructor? I can't think of a scenario when our patrol orientation would change dynamically.

File: src/software/ai/hl/stp/tactic/passer_tactic.h
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/passer_tactic.h]: Addressed in 5ae47b9

File: src/software/ai/hl/stp/tactic/chip_tactic.h
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/chip_tactic.h]: Addressed in b253531

File: src/software/ai/hl/stp/tactic/patrol_tactic.h
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/patrol_tactic.h]: Addressed in c2ad653, made it a control parameter for now. @garethellis0 @jonathanlew do you think this can be moved to the constructor as @MathewMacDougall suggested?

File: src/software/ai/hl/stp/tactic/shadow_enemy_tactic.cpp
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/shadow_enemy_tactic.cpp]: Addressed in 5186543

File: src/software/ai/hl/stp/play/shoot_or_chip_play.cpp
Comment (jsn0716) [reply src/software/ai/hl/stp/play/shoot_or_chip_play.cpp]: Addressed in 6864aef

File: src/software/ai/hl/stp/play/enemy_freekick_play.cpp
Comment (jsn0716) [reply src/software/ai/hl/stp/play/enemy_freekick_play.cpp]: Addressed in 51865437d9b10cbde18e295a7fdbd7ef6338b871

File: src/software/ai/hl/stp/tactic/tactic.h
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/tactic.h]: Addressed in 895d85b

File: src/software/ai/hl/stp/tactic/block_shot_path_tactic.h
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/block_shot_path_tactic.h]: Addressed in 5660966

File: src/software/ai/hl/stp/tactic/block_shot_path_tactic.h
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/block_shot_path_tactic.h]: Addressed in 5660966

File: src/software/ai/hl/stp/tactic/receiver_tactic.h
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/receiver_tactic.h]: Addressed in dc32704

File: src/software/ai/hl/stp/tactic/shadow_enemy_tactic.h
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/shadow_enemy_tactic.h]: Addressed in 51865437d9b10cbde18e295a7fdbd7ef6338b871

File: src/software/ai/hl/stp/tactic/shadow_enemy_tactic.h
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/shadow_enemy_tactic.h]: Addressed in feb60a45ab8b5243f2d0a31f8211a16a74181ea6 and 70a8bf4fe93a5a35f04fee4ff31010ce0a7775ea. While I was moving `enemy_team_can_pass`, I noticed that the parameter `ball_steal_speed` is never really assigned; in the constructor, the argument gets passed in but is never assigned to the parameter, and in `updateWorldParams`, it was not given as an argument so it was being assigned to itself (I removed it in 51865437d9b10cbde18e295a7fdbd7ef6338b871 since it wasn't doing anything). Should this parameter be removed from the class?

File: src/software/ai/hl/stp/tactic/shoot_goal_tactic.h
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/shoot_goal_tactic.h]: Addressed in 6864aef52b9ea29c3a441ba9a59dafcf5aff4c35

**Comment by jsn0716** [comment]
@garethellis0 Thanks for explaining the tactics and how they should be split in more detail!

File: src/software/ai/hl/stp/tactic/shadow_enemy_tactic.h
Comment (garethellis0) [reply src/software/ai/hl/stp/tactic/shadow_enemy_tactic.h]: Gut reaction is yes, but @MathewMacDougall can you confirm since you wrote this tactic?

File: src/software/ai/hl/stp/tactic/patrol_tactic.h
Comment (garethellis0) [reply src/software/ai/hl/stp/tactic/patrol_tactic.h]: Yup, I agree with @MathewMacDougall here.

**Review by garethellis0 — COMMENTED** [review]
@jsn0716 - _excellent_ job getting this turned around quickly :smile:! In terms of my comments, just the one waiting on a reply from Mathew, but otherwise this looks good to go!

File: src/software/ai/hl/stp/tactic/shadow_enemy_tactic.h
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/tactic/shadow_enemy_tactic.h]: You're right that `ball_steal_speed` wasn't assigned in the constructor, whoops :sweat_smile:. I think that it can just be put in the constructor and removed from the `update...` functions since I can't think of a use case where it would change mid-Play, but it is being used in the `calculateNextIntent` function.

**Comment by MathewMacDougall** [comment]
+1 for the quick turnaround. I replied to the one comment, and everything else LGTM.

**Review by jonathanlew — COMMENTED** [review]
Good job on the quick turnaround. @MathewMacDougall @garethellis0 What do you think about every `updateWorldParams` in each `Tactic` just taking a `World` and grabbing parameters from that, rather than each `Play` having to split up `World` into its various components. For example `CreaseDefenderTactic`

**Comment by garethellis0** [comment]
@jonathanlew - this was an explicit design decision. We don't want to just give the world to tactics because it massively increases the test surface for all the tactics. The idea was to split world params from control params so we could then write a visitor to update the world-dependent params of each tactic independent of the play.

See: https://github.com/UBC-Thunderbots/Software/issues/778

**Comment by garethellis0** [comment]
@jsn0716 - looks like you've got some conflicts

**Comment by MathewMacDougall** [comment]
@jonathanlew Also the plan is to have the `world` data passed to tactics from outside the Play. Since the tactics are returned the code in `stp.cpp` that "manages" the plays can do the updating (I need to clean up that file though...)

File: src/software/ai/hl/stp/tactic/shadow_enemy_tactic.h
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/shadow_enemy_tactic.h]: Assigning `ball_steal_speed` in the constructor seems to break `shadow_enemy_tactic_test`, so I think I want to create a separate issue to fix this.

File: src/software/ai/hl/stp/tactic/patrol_tactic.h
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/patrol_tactic.h]: Addressed in d8e9d966a2305392155236fb863a2d778e02c12a

File: src/software/ai/hl/stp/tactic/shadow_enemy_tactic.h
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/tactic/shadow_enemy_tactic.h]: Ok that's fine

**Review by MathewMacDougall — DISMISSED** [review]
LGTM :+1: Just found out I'm blocked by this so lets get it in!

**Review by garethellis0 — CHANGES_REQUESTED** [review]

**Comment by MathewMacDougall** [comment]
Oh we got a conflict again somehow

File: src/software/ai/hl/stp/tactic/shadow_enemy_tactic.h
Comment (jsn0716) [reply src/software/ai/hl/stp/tactic/shadow_enemy_tactic.h]: My mistake, fixed it in 3b5a70da34710640ea5e25c440eeeaeb28c7e6f0

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada: - thanks for your persistence in fixing things up to get this in!

**Review by MathewMacDougall — APPROVED** [review]
:+1: :tada:

**Review by MathewMacDougall — APPROVED** [review]
:+1: :tada:

**Comment by MathewMacDougall** [comment]
@akhilveeraghanta @jonathanlew need your approvals to merge.

**Review by akhilveeraghanta — APPROVED** [review]
Lgtm nice work!

**Review by jonathanlew — APPROVED** [review]
Thanks for the fast turnaround

**Pull request merged**
